GVFS/GVFS/RepairJobs/GitRefsRepairJob.cs

```diff
@@ -0,0 +1,163 @@
+        /// to inspect for and correct issues.
+        /// </summary>
+        /// <returns></returns>
+        protected abstract IEnumerable<string> GetRefs();
+
+        /// <summary>
+        /// Check if the contents of a reference is valid.
+        /// </summary>
+        /// <param name="fullSymbolicRef">A full symbolic ref name. eg. HEAD, refs/remotes/origin/HEAD, refs/heads/master</param>
+        /// <param name="refContents">Contents of the ref file</param>
+        protected virtual bool IsValidRefContents(string fullSymbolicRef, string refContents)
+        {
+            return SHA1Util.IsValidShaFormat(refContents);
+        }
+
+        /// <summary>
+        /// 'git ref-log' doesn't work if the repo is corrupted, so parsing reflogs seems like the only solution.
+        /// </summary>
+        /// <param name="fullSymbolicRef">A full symbolic ref name. eg. HEAD, refs/remotes/origin/HEAD, refs/heads/master</param>
+        private static bool TryReadLastRefLogEntry(Enlistment enlistment, string fullSymbolicRef, out RefLogEntry refLog, out string error)
+        {
+            string refLogPath = Path.Combine(enlistment.WorkingDirectoryRoot, GVFSConstants.DotGit.Logs.Root, fullSymbolicRef);
+            if (!File.Exists(refLogPath))
+            {
+                refLog = null;
+                error = $"Could not find reflog for ref '{fullSymbolicRef}'";
+                return false;
+            }
+
+            try
+            {
+                string refLogContents = File.ReadLines(refLogPath).Last();
+                if (!RefLogEntry.TryParse(refLogContents, out refLog))
+                {
+                    error = $"Last ref log entry for '{fullSymbolicRef}' is unparsable.";
+                    return false;
+                }
+            }
+            catch (IOException ex)
+            {
+                refLog = null;
+                error = $"IOException while reading reflog '{fullSymbolicRef}': " + ex.Message;
+                return false;
+            }
+
+            error = null;
+            return true;
+        }
+
+        private bool TryParseRef(string fullSymbolicRef, List<string> messages = null)
+        {
+            string refPath = Path.Combine(this.Enlistment.WorkingDirectoryRoot, GVFSConstants.DotGit.Root, fullSymbolicRef);
+            if (!File.Exists(refPath))
+            {
+                messages?.Add($"Could not find ref file for '{fullSymbolicRef}'");
+                return false;
+            }
+
+            string refContents;
+            try
+            {
+                refContents = File.ReadAllText(refPath).Trim();
+            }
+            catch (IOException ex)
+            {
+                messages?.Add($"IOException while reading '{refPath}': {ex.Message}");
+                return false;
+            }
+
+            if (this.IsValidRefContents(fullSymbolicRef, refContents))

```

Comment:
How are we resolving symbolic refs that have a symbolic ref?  For example files in `refs/heads/`  can have a ref in them and don't have to be a sha by running `git symbolic-ref refs/heads/branch2 refs/heads/branch1`

Reply:
I pushed the symbolic ref check down into the base.